File: sdk/python/src/openagentio/bus/stream.py
"""Stream framing — server-side StreamWriter and client-side Stream iterator.

Mirrors pkg/bus/stream.go. A streaming response is a sequence of envelopes
sharing the same ``correlation_id``; the runtime tags each frame with a
monotonic ``seq`` and a terminal frame (``ResponseFinal`` or ``ResponseError``)
sets ``is_final = True``.
"""
from __future__ import annotations

import asyncio
from typing import Any

from openagentio.codec.json_codec import Codec
from openagentio.event.envelope import Envelope
from openagentio.event.payload import CodeAgentUnavailable, ErrorPayload
from openagentio.event.types import (
    ResponseDelta,
    ResponseError,
    ResponseFinal,
    ResponseStarted,
)
from openagentio.transport.base import Inbox, RawMessage, Transport


class ErrIdleTimeout(Exception):
    """Raised when the gap between two streaming frames exceeds the idle timeout."""


class StreamWriter:
    """Server-side stream emitter.

    Each method publishes one frame back to the request's ``reply_to`` subject.
    ``started`` may be called at most once. ``final`` and ``error`` are
    mutually exclusive and terminal — calls after either raise.
    """

    def __init__(
        self,
        transport: Transport,
        codec: Codec,
        agent_id: str,
        request: Envelope,
    ) -> None:
        self._transport = transport
        self._codec = codec
        self._agent_id = agent_id
        self._req = request
        self._lock = asyncio.Lock()
        self._seq = 0
        self._started = False
        self._closed = False

    @property
    def closed(self) -> bool:
        return self._closed

    async def started(self, meta: Any = None) -> None:
        async with self._lock:
            if self._closed:
                raise RuntimeError("stream: already closed")
            if self._started:
                raise RuntimeError("stream: started already emitted")
            self._started = True
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseStarted)
        env.seq = seq
        env.payload = self._codec.encode_payload(meta)
        await self._publish(env)

    async def delta(self, chunk: Any = None) -> None:
        async with self._lock:
            if self._closed:
                raise RuntimeError("stream: already closed")
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseDelta)
        env.seq = seq
        env.payload = self._codec.encode_payload(chunk)
        await self._publish(env)

    async def final(self, result: Any = None) -> None:
        async with self._lock:
            if self._closed:
                raise RuntimeError("stream: already closed")
            self._closed = True
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseFinal)
        env.seq = seq
        env.is_final = True
        env.payload = self._codec.encode_payload(result)
        await self._publish(env)

    async def error(self, exc: BaseException) -> None:
        async with self._lock:
            if self._closed:
                raise RuntimeError("stream: already closed")
            self._closed = True
            seq = self._next_seq_locked()

        env = new_reply_shell(self._agent_id, self._req, ResponseError)
        env.seq = seq
        env.is_final = True
        payload = ErrorPayload(code=CodeAgentUnavailable, message=str(exc))
        env.payload = self._codec.encode_payload(payload)
        await self._publish(env)

    def _next_seq_locked(self) -> int:
        s = self._seq
        self._seq += 1
        return s

    async def _publish(self, env: Envelope) -> None:
        data = self._codec.encode_envelope(env)
        await self._transport.publish(
            RawMessage(subject=self._req.reply_to, data=data)
        )


class Stream:
    """Async iterator over a streaming response.

    Frames are reordered by :py:attr:`Envelope.seq`; the iterator stops after
    yielding a frame with ``is_final = True``.

    Two timeouts coexist (mirroring the Go SDK):

    * ``idle_timeout`` — maximum gap between two frames; expiring raises
      :class:`ErrIdleTimeout`.
    * ``deadline`` — absolute wall-clock deadline for the whole stream
      (in :py:meth:`asyncio.AbstractEventLoop.time` units); expiring raises
      :class:`asyncio.TimeoutError`. The overall deadline always wins when
      both timers would fire — the iterator checks the deadline before and
      after each ``recv`` call.
    """

    def __init__(
        self,
        inbox: Inbox,
        codec: Codec,
        idle_timeout: float | None = None,
        deadline: float | None = None,
    ) -> None:
        self._inbox = inbox
        self._codec = codec
        self._idle = idle_timeout
        self._deadline = deadline
        self._expected = 0
        self._pending: dict[int, Envelope] = {}
        self._exhausted = False

    def __aiter__(self) -> "Stream":
        return self

    async def __anext__(self) -> Envelope:
        if self._exhausted:
            raise StopAsyncIteration

        while True:
            ready = self._pending.pop(self._expected, None)
            if ready is not None:
                self._expected += 1
                if ready.is_final:
                    self._exhausted = True
                return ready

            wait = self._idle
            if self._deadline is not None:
                remaining = self._deadline - asyncio.get_running_loop().time()
                if remaining <= 0:
                    self._exhausted = True
                    raise asyncio.TimeoutError("bus: stream overall timeout")
                wait = remaining if wait is None else min(wait, remaining)

            try:
                msg = await self._inbox.recv(timeout=wait)
            except asyncio.TimeoutError:
                self._exhausted = True
                if (
                    self._deadline is not None
                    and asyncio.get_running_loop().time() >= self._deadline
                ):
                    raise asyncio.TimeoutError(
                        "bus: stream overall timeout"
                    ) from None
                raise ErrIdleTimeout("bus: stream idle timeout") from None
            except RuntimeError:
                # inbox closed mid-iteration — clean exit.
                self._exhausted = True
                raise StopAsyncIteration from None

            env = self._codec.decode_envelope(msg.data)
            if env.seq < self._expected:
                continue  # late / duplicate frame
            if env.seq in self._pending:
                continue
            self._pending[env.seq] = env

    async def close(self) -> None:
        if self._exhausted:
            return
        self._exhausted = True
        try:
            await self._inbox.close()
        except Exception:
            pass


def new_reply_shell(agent_id: str, req: Envelope, event_type: str) -> Envelope:
    """Pre-populate a response envelope with correlation metadata copied from req."""
    resp = Envelope.new(event_type)
    resp.from_ = agent_id
    resp.to = req.from_
    resp.session_id = req.session_id
    resp.conversation_id = req.conversation_id
    resp.tenant_id = req.tenant_id
    resp.user_id = req.user_id
    resp.channel = req.channel
    resp.trace_id = req.trace_id
    resp.span_id = req.span_id
    resp.traceparent = req.traceparent
    resp.correlation_id = req.event_id
    return resp

File: pkg/bus/stream.go
package bus

import (
	"context"
	"errors"
	"iter"
	"sync"
	"time"

	"github.com/ModulationAI/openagentio/pkg/codec"
	"github.com/ModulationAI/openagentio/pkg/event"
	"github.com/ModulationAI/openagentio/pkg/middleware"
	"github.com/ModulationAI/openagentio/pkg/transport"
)

// StreamInvoke publishes a request to {prefix}.invoke.{target} with a fresh
// _INBOX as reply_to and returns a Stream over the received frames. Frames
// are reordered by Envelope.Seq; the iterator terminates after a frame with
// IsFinal=true, or earlier on idle/overall timeout.
func (b *defaultBus) StreamInvoke(ctx context.Context, target string, payload any, opts ...InvokeOption) (Stream, error) {
	if target == "" {
		return nil, errors.New("bus: empty invoke target")
	}
	o := collectInvokeOpts(opts)
	timeout := o.Timeout
	if timeout == 0 {
		timeout = b.opts.DefaultTimeout
	}

	streamCtx, cancel := context.WithCancel(ctx)
	if timeout > 0 {
		streamCtx, cancel = context.WithTimeout(ctx, timeout)
	}

	env, err := b.buildRequestEnvelope(target, payload)
	if err != nil {
		cancel()
		return nil, err
	}

	inbox, err := b.opts.Transport.OpenInbox(streamCtx)
	if err != nil {
		cancel()
		return nil, err
	}
	env.ReplyTo = inbox.Subject()

	b.prepareEnvelope(streamCtx, env)
	data, err := b.opts.Codec.EncodeEnvelope(env)
	if err != nil {
		_ = inbox.Close()
		cancel()
		return nil, err
	}
	if err := b.opts.Transport.Publish(streamCtx, &transport.RawMessage{
		Subject: b.invokeSubject(target, b.resolveTenant(env.TenantID)),
		Data:    data,
	}); err != nil {
		_ = inbox.Close()
		cancel()
		return nil, err
	}

	return &stream{
		ctx:    streamCtx,
		cancel: cancel,
		inbox:  inbox,
		codec:  b.opts.Codec,
		idle:   o.IdleTimeout,
	}, nil
}

// HandleStream subscribes to {prefix}.invoke.{target} and dispatches each
// request into a goroutine running the supplied handler. A StreamWriter is
// provided that publishes started/delta/final/error frames back to
// req.ReplyTo with monotonically increasing Seq numbers. If the handler
// returns without calling Final or Error, the runtime auto-emits one based on
// the returned error.
func (b *defaultBus) HandleStream(target string, h StreamHandler, opts ...HandleOption) error {
	if target == "" {
		return errors.New("bus: empty invoke target")
	}
	if h == nil {
		return errors.New("bus: nil stream handler")
	}
	o := collectHandleOpts(opts)
	if !o.QueueSet {
		o.Queue = target
	}
	subject := b.invokeSubject(target, b.opts.Tenant)

	dispatch := func(_ context.Context, msg *transport.RawMessage) error {
		req, err := b.opts.Codec.DecodeEnvelope(msg.Data)
		if err != nil {
			return err
		}
		if req.ReplyTo == "" {
			return errors.New("bus: stream request missing reply_to")
		}
		go b.handleStream(req, h)
		return nil
	}

	sub, err := b.opts.Transport.Subscribe(b.lifeCtx, subject, o.Queue, dispatch)
	if err != nil {
		return err
	}
	b.trackOwned(sub)
	return nil
}

func (b *defaultBus) handleStream(req *event.Envelope, h StreamHandler) {
	ctx, cancel := context.WithCancel(b.lifeCtx)
	defer cancel()

	w := &streamWriter{
		bus: b,
		ctx: ctx,
		req: req,
	}

	chained := middleware.Chain(middleware.Handler(func(c context.Context, e *event.Envelope) error {
		return h(c, e, w)
	}), b.opts.Middleware...)

	herr := chained(ctx, req)

	if w.isClosed() {
		return
	}
	if herr != nil {
		_ = w.Error(herr)
	} else {
		_ = w.Final(nil)
	}
}

// --- client-side stream ------------------------------------------------------

type stream struct {
	ctx    context.Context
	cancel context.CancelFunc
	inbox  transport.Inbox
	codec  codec.Codec
	idle   time.Duration

	closeOnce sync.Once
	closeErr  error
}

func (s *stream) Close() error {
	s.closeOnce.Do(func() {
		s.cancel()
		s.closeErr = s.inbox.Close()
	})
	return s.closeErr
}

func (s *stream) Events() iter.Seq2[*event.Envelope, error] {
	return func(yield func(*event.Envelope, error) bool) {
		var (
			expected uint64
			pending  = make(map[uint64]*event.Envelope)
		)

		for {
			recvCtx := s.ctx
			var recvCancel context.CancelFunc
			if s.idle > 0 {
				recvCtx, recvCancel = context.WithTimeout(s.ctx, s.idle)
			}
			msg, err := s.inbox.Recv(recvCtx)
			if recvCancel != nil {
				recvCancel()
			}

			if err != nil {
				switch {
				case s.ctx.Err() != nil:
					yield(nil, s.ctx.Err())
				case s.idle > 0 && errors.Is(err, context.DeadlineExceeded):
					yield(nil, ErrIdleTimeout)
				default:
					yield(nil, err)
				}
				return
			}

			env, err := s.codec.DecodeEnvelope(msg.Data)
			if err != nil {
				yield(nil, err)
				return
			}

			if env.Seq < expected {
				continue // duplicate / late frame, drop
			}
			if _, dup := pending[env.Seq]; dup {
				continue
			}
			pending[env.Seq] = env

			for {
				e, ok := pending[expected]
				if !ok {
					break
				}
				delete(pending, expected)
				expected++
				if !yield(e, nil) {
					return
				}
				if e.IsFinal {
					return
				}
			}
		}
	}
}

// --- server-side writer ------------------------------------------------------

type streamWriter struct {
	bus *defaultBus
	ctx context.Context
	req *event.Envelope

	mu      sync.Mutex
	seq     uint64
	started bool
	closed  bool
}

func (w *streamWriter) Started(meta any) error {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return errors.New("stream: already closed")
	}
	if w.started {
		w.mu.Unlock()
		return errors.New("stream: started already emitted")
	}
	w.started = true
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseStarted)
	env.Seq = seq
	if meta != nil {
		data, err := w.bus.opts.Codec.EncodePayload(meta)
		if err != nil {
			return err
		}
		env.Payload = data
	}
	return w.publish(env)
}

func (w *streamWriter) Delta(chunk any) error {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return errors.New("stream: already closed")
	}
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseDelta)
	env.Seq = seq
	if chunk != nil {
		data, err := w.bus.opts.Codec.EncodePayload(chunk)
		if err != nil {
			return err
		}
		env.Payload = data
	}
	return w.publish(env)
}

func (w *streamWriter) Final(result any) error {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return errors.New("stream: already closed")
	}
	w.closed = true
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseFinal)
	env.Seq = seq
	env.IsFinal = true
	if result != nil {
		data, err := w.bus.opts.Codec.EncodePayload(result)
		if err != nil {
			return err
		}
		env.Payload = data
	}
	return w.publish(env)
}

func (w *streamWriter) Error(srcErr error) error {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return errors.New("stream: already closed")
	}
	w.closed = true
	seq := w.nextSeqLocked()
	w.mu.Unlock()

	env := newReplyShell(w.bus.opts.AgentID, w.req, event.ResponseError)
	env.Seq = seq
	env.IsFinal = true
	payload := event.ErrorPayload{
		Code:    event.CodeAgentUnavailable,
		Message: srcErr.Error(),
	}
	data, _ := w.bus.opts.Codec.EncodePayload(payload)
	env.Payload = data
	return w.publish(env)
}

func (w *streamWriter) isClosed() bool {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.closed
}

func (w *streamWriter) nextSeqLocked() uint64 {
	s := w.seq
	w.seq++
	return s
}

func (w *streamWriter) publish(env *event.Envelope) error {
	data, err := w.bus.opts.Codec.EncodeEnvelope(env)
	if err != nil {
		return err
	}
	return w.bus.opts.Transport.Publish(w.ctx, &transport.RawMessage{
		Subject: w.req.ReplyTo,
		Data:    data,
	})
}
